group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -285,6 +302,64 @@ public ConsumerGroupMember getOrMaybeCreateMember(
         return member;
     }
 
+    /**
+     * Gets or creates a static member.
+     *
+     * @param memberId          The member id.
+     * @param instanceId        The group instance id.
+     * @param createIfNotExists Booleans indicating whether the member must be
+     *                          created if it does not exist.
+     *
+     * @return A ConsumerGroupMember.
+     */
+    public ConsumerGroupMember getOrMaybeCreateStaticMember(
+            String memberId,
+            String instanceId,
+            boolean createIfNotExists
+    ) {
+        ConsumerGroupMember member;
+        String existingMemberId = staticMemberId(instanceId);
+        if (!createIfNotExists) {
+            // The member joined with a non-zero epoch but we haven't registered this static member
+            // This could be an unknown member for the coordinator.
+            if (existingMemberId == null) {
+                throw Errors.UNKNOWN_MEMBER_ID.exception();
+            }
+            // We can't create a member at this point. If the 2 member-ids don't match,
+            // we will throw an error.
+            if (!existingMemberId.equals(memberId)) {
+                throw Errors.FENCED_INSTANCE_ID.exception();
+            }
+            member = getOrMaybeCreateMember(memberId, false);
+        } else {
+            // No existing member found against this instance id. Creating new.
+            if (existingMemberId == null) {
+                member = getOrMaybeCreateMember(memberId, true);
+                staticMembers.put(instanceId, memberId);
+                return member;
+            } else {
+                // Get the details of the existing member
+                ConsumerGroupMember existingMember = getOrMaybeCreateMember(existingMemberId, false);
+                int currentMemberEpoch = existingMember.memberEpoch();
+                // A new member with a used instance id joined but the previous member using the same instance id
+                // hasn't requested leaving the group.
+                if (currentMemberEpoch != -2 && !existingMemberId.equals(memberId)) {
+                    throw Errors.UNRELEASED_INSTANCE_ID.exception();
+                }
+                // A new static member is trying to take the place of a departed static member. We will
+                // provide the assignments of the old member to the new one.
+                member = new ConsumerGroupMember.Builder(memberId, existingMember)
+                        .setMemberEpoch(existingMember.targetMemberEpoch())
+                        .setPreviousMemberEpoch(0)
+                        .setTargetMemberEpoch(existingMember.targetMemberEpoch())
+                        .build();
```

Comment:
This is a bit weird because you pass `existingMember` to the builder and then you still have to override other fields. Would it be better to do `new ConsumerGroupMember.Builder(existingMember)` and then override the fields? I think that we only need to set the new member id.

nit: The indentation should be four spaces.

Reply:
YEs, that didn't quite make sense. It's not needed anymore. Fixed indent as well.